File: examples/Reduce.hs
-- ---------
-- Reduce.hs
-- ---------

{-
(>>)     :: IO a          -> IO b      -> IO b -- 'then' operator
(+)      :: a             -> a         -> a
(*)      :: a             -> a         -> a
assert   :: Bool          -> a         -> a
foldl    :: (a -> b -> a) -> a  -> [b] -> a
putStrLn :: String                     -> IO ()
return   :: a                          -> IO a
-}

import Control.Exception (assert)

reduce :: (a -> a -> a) -> a -> [a] -> a
reduce _  v []       = v
reduce bf v (x : xs) = reduce bf (bf v x) xs

test :: ((Int -> Int -> Int) -> Int -> [Int] -> Int) -> IO ()
test f =
    let a :: [Int]
        a = [2, 3, 4]
    in assert ((f (+) 0 a) ==  9) return () >>
       assert ((f (*) 1 a) == 24) return ()

main :: IO ()
main =
    putStrLn "Reduce.hs" >>

    test foldl  >>
    test reduce >>

    putStrLn "Done."

File: examples/Reduce.py
#!/usr/bin/env python

# ---------
# Reduce.py
# ---------

import operator

def reduce1 (bf, a, v) :
    if not a :
        return v
    return reduce1(bf, a[1:], bf(v, a[0]))

def reduce2 (bf, a, v) :
    i = 0
    s = len(a)
    while i != s :
        v = bf(v, a[i])
        i += 1
    return v

def reduce3 (bf, a, v) :
    i = iter(a)
    while True :
        try :
            v = bf(v, i.next())
        except StopIteration :
            break
    return v

def reduce4 (bf, a, v) :
    for w in a :
        v = bf(v, w)
    return v

def test (f) :
    a = [2, 3, 4]
    assert f(operator.add, a, 0) ==  9
    assert f(operator.mul, a, 1) == 24

print "Reduce.py"

test(reduce)
test(reduce1)
test(reduce2)
test(reduce3)
test(reduce4)

print "Done."
